File: 0145/recursive.cpp
#include <cassert>
#include <iostream>
#include <vector>

#include "leetcode/tree.hpp"

using namespace std;
using namespace LeetCode;

class Solution {
 public:
  vector<int> postorderTraversal(TreeNode* root) {
    vector<int> ret;
    dfs(root, ret);
    return ret;
  }

  void dfs(TreeNode* node, vector<int>& ret) {
    if (!node) return;
    dfs(node->left, ret);
    dfs(node->right, ret);
    ret.push_back(node->val);
  }
};

int main() {
  Solution sol;
  string str;
  TreeNode* root;
  vector<int> got, expected;

  root = CreateTreeNode("1 # 2 # # 3 ;");
  expected = {3,2,1};
  got = sol.postorderTraversal(root);
  assert(got == expected);
  DeleteTreeNode(root);

  root = CreateTreeNode("1 2 3 4 5 6 7 ;");
  expected = {4,5,2,6,7,3,1};
  got = sol.postorderTraversal(root);
  assert(got == expected);
  DeleteTreeNode(root);

  got = sol.postorderTraversal(nullptr);
  assert(got.size() == 0);
}

File: 0107/recursive.py
from typing import *


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        ret = []
        self.dfs(root, 0, ret)
        return ret

    def dfs(self, node, level, ret):
        if node is None:
            return
        if level >= len(ret):
            ret.insert(0, [node.val])
        else:
            ret[-1-level].append(node.val)
        self.dfs(node.left, level+1, ret)
        self.dfs(node.right, level+1, ret)


if __name__ == '__main__':
    sol = Solution()

    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    assert sol.levelOrderBottom(root) == [[15,7],[9,20],[3]]

    assert sol.levelOrderBottom(None) == []
